app/forms/validators.py

```diff
@@ -225,6 +225,14 @@ def __call__(self, form, field):
 
         if not form.data or not re.match(r"\d{4}$", str(form.year.data)):
             raise validators.StopValidation(self.message)
+        if re.match(r"^\d{4}-", str(form.data)) and not re.match(
+            r"\d{1,2}", str(form.month.data)
+        ):
+            raise validators.StopValidation(self.message)
+        if re.match(r"^\d{4}-\d{1,2}-", str(form.data)) and not re.match(
+            r"\d{1,2}", str(form.day.data)
+        ):
+            raise validators.StopValidation(self.message)
```

Comment:
I don't think we need to check `form.data` each time, so this can be simplified to:
```python
        if (
            not form.data
            or not re.match(r"\d{4}$", str(form.year.data))
            or not re.match(r"\d{1,2}", str(form.month.data))
            or not re.match(r"\d{1,2}", str(form.day.data))
        ):
            raise validators.StopValidation(self.message)
```

Reply:
Some of the forms we use don't have `day` and/or `month` attributes and it breaks runner this way (when looking for `form.month.data` and `form.day.data`).

Reply:
Ah ok. In that case we should explicitly check for those attributes, something like:
```python
        if (
            not form.data
            or not re.match(r"\d{4}$", str(form.year.data))
            or (hasattr(form, "month") and not re.match(r"\d{1,2}", str(form.month.data)))
            or (hasattr(form, "day") and not re.match(r"\d{1,2}", str(form.day.data)))
        ):
            raise validators.StopValidation(self.message)
```

Reply:
Correct me if I am wrong but could we not replace this whole method with:
```py
        if not form.data or not hasattr(form, "year"):
            raise validators.StopValidation(self.message)

        try:
            if hasattr(form, "day"):
                datetime.strptime(form.data, "%Y-%m-%d")
            elif hasattr(form, "month"):
                datetime.strptime(form.data, "%Y-%m")
            else:
                datetime.strptime(form.data, "%Y")
        except ValueError:
            raise validators.StopValidation(self.message)
```

Reply:
Refactored, so it's using this solution above now, thanks👍.

Reply:
Nice 👌